server/application/application.go

```diff
@@ -456,19 +456,25 @@ func (s *Server) Rollback(ctx context.Context, rollbackReq *ApplicationRollbackR
 }
 
 func (s *Server) setAppOperation(ctx context.Context, appName string, operationCreator func(app *appv1.Application) (*appv1.Operation, error)) (*appv1.Application, error) {
-	app, err := s.Get(ctx, &ApplicationQuery{Name: &appName})
-	if err != nil {
-		return nil, err
-	}
-	if app.Operation != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "another operation is already in progress")
-	}
-	op, err := operationCreator(app)
-	if err != nil {
-		return nil, err
+	for {
+		a, err := s.Get(ctx, &ApplicationQuery{Name: &appName})
```

Comment:
Looks good!  One small concern: I don't know how many loops this is expected to run on average.  Depending on the answer to that, would there be any advantage to creating the `ApplicationQuery` struct before the `for` loop begins, rather than recreating it each time through?

Reply:
I think it will run 1~3 times max. This is possible only if app controller is actively updating app in background. In this case moving `ApplicationQuery` creation outside of loop would not give big advantage.